core/src/main/java/org/apache/spark/shuffle/api/ShuffleDataIO.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.api;
+
+import org.apache.spark.annotation.Private;
+
+/**
+ * :: Experimental ::
+ * An interface for plugging in modules for storing and reading temporary shuffle data.
+ * <p>
+ * A single instance of this module is loaded per process in the Spark application.
+ * The default implementation reads and writes shuffle data from the local disks of
+ * the executor, and is the implementation of shuffle file storage that has remained
+ * consistent throughout most of Spark's history.
+ * <p>
+ * Alternative implementations of shuffle data storage can be loaded via setting
+ * spark.shuffle.io.plugin.class.
+ * @since 3.0.0
+ */
+@Private
+public interface ShuffleDataIO {
```

Comment:
I think naming it to `IO` is great, because it means this is a more underground pluggable layer than `ShuffleManager`, but can all readers/developers get this? Should we explain this naming and its difference with `ShuffleManager`?

Reply:
I'll add some JavaDoc explaining the difference between the ShuffleManager plugin and this plugin system.

Reply:
A question that may be naive, why do we choose Java over Scala? I see Spark classes except the ones dealing with underlying memory write in Scala...

Reply:
As a public interface, it is better to use Java, so that other users can implement with Java, Scala  or other jvm languages. 

If we defined the APIs using Scala, mostly user can only use Scala to implement it, unless it is well designed to avoid Scala specific features, so that it can be leveraged by Java.